src/cargo/ops/cargo_rustc/mod.rs

```diff
@@ -703,23 +700,34 @@ fn rustdoc<'a, 'cfg>(cx: &mut Context<'a, 'cfg>,
 }
 
 // The path that we pass to rustc is actually fairly important because it will
-// show up in error messages and the like. For this reason we take a few moments
-// to ensure that something shows up pretty reasonably.
+// show up in error messages (important for readability), debug information
+// (important for caching), etc. As a result we need to be pretty careful how we
+// actually invoke rustc.
 //
-// The heuristic here is fairly simple, but the key idea is that the path is
-// always "relative" to the current directory in order to be found easily. The
-// path is only actually relative if the current directory is an ancestor if it.
-// This means that non-path dependencies (git/registry) will likely be shown as
-// absolute paths instead of relative paths.
-fn root_path(cx: &Context, unit: &Unit) -> PathBuf {
-    let absolute = unit.pkg.root().join(unit.target.src_path());
-    let cwd = cx.config.cwd();
-    if absolute.starts_with(cwd) {
-        util::without_prefix(&absolute, cwd).map(|s| {
-            s.to_path_buf()
-        }).unwrap_or(absolute)
-    } else {
-        absolute
+// In general users don't expect `cargo build` to cause rebuilds if you change
+// directories. That could be if you just change directories in the project or
+// if you literally move the whole project wholesale to a new directory. As a
+// result we mostly don't factor in `cwd` to this calculation. Instead we try to
+// track the workspace as much as possible and we update the current directory
+// of rustc/rustdoc where approrpriate.
+//
+// The first returned value here is the argument to pass to rustc, and the
+// second is the cwd that rustc should operate in.
+fn path_args(cx: &Context, unit: &Unit) -> (PathBuf, Option<PathBuf>) {
+    let ws_root = cx.ws.root();
+    let src = unit.target.src_path();
+    assert!(src.is_absolute());
+    match src.strip_prefix(ws_root) {
+        Ok(path) => (path.to_path_buf(), Some(ws_root.to_path_buf())),
+        Err(_) => (src.to_path_buf(), None),
```

Comment:
Returning `None` here means that `rustc` will use pretty-much arbitrary cwd, is that right? Maybe it's better to lock this down to `ws_root` even in this case, just to get a bit more determinism? Than we can return just `PathBuf` here and avoid that comment about hashing only `.0`.

Reply:
Certainly plausible! I was originally thinking that this wouldn't matter but I think with plugins nowadays it could matter for sure. I'll change.

That being said though I don't think we can avoid hashing the second field of the return value because `ws_root` changes over time (if the whole dir is renamed) but we don't want that to cause recompiles.

Reply:
>That being said though I don't think we can avoid hashing the second field of the return value because ws_root changes over time (if the whole dir is renamed) but we don't want that to cause recompiles.

The idea is that if we lock down cwd, then we don't need to return a pair here, we can return just the path.